mmf/trainers/lightning_core/loop_callback.py

```diff
@@ -47,3 +92,134 @@ def on_validation_batch_end(
     ):
         # prepare the next batch
         self.lightning_trainer.data_module.val_loader.change_dataloader()
+
+        # aggregate val_combined_report
+        self.val_combined_report = self._update_and_create_report(
+            outputs, batch_idx, self.val_combined_report, update_meter=self.val_meter
+        )
+        self.val_combined_report.metrics = self.lightning_trainer.metrics(
+            self.val_combined_report, self.val_combined_report
+        )
+        self.update_meter(self.val_combined_report, self.val_meter, eval_mode=True)
+
+    def on_validation_end(self, trainer: Trainer, pl_module: LightningModule):
+        iterations = self._get_iterations_for_logging(trainer)
+        current_epochs = self._get_current_epoch_for_logging(trainer)
+        num_updates = self._get_num_updates_for_logging(trainer)
+        extra = {
+            "num_updates": num_updates,
+            "epoch": current_epochs,
+            "iterations": iterations,
+            "max_updates": trainer.max_steps,
+            "val_time": self.snapshot_timer.get_time_since_start(),
+        }
+        # TODO: @sash populate early stop info for logging (next mvp)
+        # extra.update(self.trainer.early_stop_callback.early_stopping.get_info())
+        self.train_timer.reset()
+        summarize_report(
+            current_iteration=iterations,
+            num_updates=num_updates,
+            max_updates=trainer.max_steps,
+            meter=self.val_meter,
+            extra=extra,
+            tb_writer=self.lightning_trainer.tb_writer,
+        )
+
+    def _update_and_create_report(
+        self,
+        step_output: Dict,
+        batch_idx: int,
+        combined_report: Report = None,
+        update_meter: Meter = None,
+    ):
+        input_batch = step_output["input_batch"]
+        report = Report(input_batch, step_output)
+
+        if update_meter:
+            self.update_meter(report, update_meter)
+
+        should_accumulate = not (
+            batch_idx % self.trainer_config.accumulate_grad_batches == 0
+        )
+
+        final_report = report
+        if should_accumulate and combined_report is not None:
+            combined_report.accumulate_tensor_fields_and_loss(
+                report, self.lightning_trainer.metrics.required_params
+            )
+            combined_report.batch_size += report.batch_size
+            final_report = combined_report
+
+        return final_report
+
+    def get_optimizer(self, trainer: Trainer):
+        assert (
+            len(trainer.optimizers) == 1
+        ), "mmf lightning_trainer supports 1 optimizer per model for now."
+        optimizer = trainer.optimizers[0]
+        return optimizer
+
+    def _save_checkpoint(self, trainer: Trainer):
+        logger.info("Checkpoint time. Saving a checkpoint.")
+        return
+        # TODO: sash Needs implementation - next mvp
+
+    def _get_current_epoch_for_logging(self, trainer: Trainer):
```

Comment:
Shouldn't these be properties?

Reply:
I do think properties is the best - since it is just + 1 from the original number.